system/include/emscripten/fetch.h

```diff
@@ -0,0 +1,179 @@
+	//     body of the downloaded data. Otherwise this will be null.
+	// In onprogress() handler:
+	//   - If the EMSCRIPTEN_FETCH_STREAM_DATA attribute was specified for the transfer, this points to a partial
+	//     chunk of bytes related to the transfer. Otherwise this will be null.
+	const char *data; // XXX 12
+
+	// Specifies the length of the above data block in bytes. When the download finishes, this field will be valid even if
+	// EMSCRIPTEN_FETCH_LOAD_TO_MEMORY was not specified.
+	uint64_t numBytes; // XXX 16
+
+	// If EMSCRIPTEN_FETCH_STREAM_DATA is being performed, this indicates the byte offset from the start of the stream
+	// that the data block specifies. (for onprogress() streaming XHR transfer, the number of bytes downloaded so far before this chunk)
+	uint64_t dataOffset; // XXX 24
+
+	// Specifies the total number of bytes that the response body will be.
+	uint64_t totalBytes; // XXX 32
+
+	// Specifies the readyState of the XHR request:
+	// 0: UNSENT: request not sent yet
+	// 1: OPENED: emscripten_fetch has been called.
+	// 2: HEADERS_RECEIVED: emscripten_fetch has been called, and headers and status are available.
+	// 3: LOADING: download in progress.
+	// 4: DONE: download finished.
+	// See https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/readyState
+	unsigned short readyState; // XXX 40
+
+	// Specifies the status code of the response.
+	unsigned short status; // XXX 42
+
+	// Specifies a human-readable form of the status code.
+	char statusText[64]; // XXX 44
+
+	uint32_t __proxyState; // XXX 108
+
+	// For internal use only.
+	emscripten_fetch_attr_t __attributes; // XXX 112
+};
+
+// Clears the fields of an emscripten_fetch_attr_t structure to their default values in a future-compatible manner.
+void emscripten_fetch_attr_init(emscripten_fetch_attr_t *fetch_attr);
+
+// Initiates a new Emscripten fetch operation, which downloads data from the given URL or from IndexedDB database.
+emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const char *url);
+
+// Synchronously blocks to wait for the given fetch operation to complete. This operation is not allowed in the main browser
+// thread, in which case it will return EMSCRIPTEN_RESULT_NOT_SUPPORTED. Pass timeoutMSecs=infinite to wait indefinitely. If
+// the wait times out, the return value will be EMSCRIPTEN_RESULT_TIMEOUT.
+// The onsuccess()/onerror()/onprogress() handlers will be called in the calling thread from within this function before
+// this function returns.
+EMSCRIPTEN_RESULT emscripten_fetch_wait(emscripten_fetch_t *fetch, double timeoutMSecs);
+
+// Closes a finished or an executing fetch operation and frees up all memory. If the fetch operation was still executing, the
+// onerror() handler will be called in the calling thread before this function returns.
+EMSCRIPTEN_RESULT emscripten_fetch_close(emscripten_fetch_t *fetch);
```

Comment:
That was currently unimplemented, `emscripten_fetch_close` will free `data` and call `onerror` if the close caused a fetch to abort mid-flight, so the intent is that the function will behave like documented.